### ISSUE TITLE
File save blocking causes hot reload delays compared to other editors

### ISSUE THREAD

**Issue by sausaw** [zed-industries/zed#34527]
# File Save Blocking Causes Hot Reload Delays

## Summary

Hot reload systems (like Vite, Next.js dev server, etc.) pick up file changes significantly later when saving files from Zed compared to other editors like VS Code or Sublime Text. This is caused by Zed's file save operation blocking on filesystem entry refresh, delaying the save completion signal that hot reload tools rely on.

## Root Cause

The issue is in `crates/worktree/src/worktree.rs` in the `LocalWorktree::write_file` implementation around line 1667:

```rust
cx.spawn(async move |this, cx| {
    write.await?;  // ← File is written to disk here
    let entry = this
        .update(cx, |this, cx| {
            this.as_local_mut()
                .unwrap()
                .refresh_entry(path.clone(), None, cx)  // ← This blocks!
        })?
        .await?;  // ← Save completion waits for this
    // ... rest of save logic
})
```

### The Problem Flow

1. **File is written to disk** (`write.await?`) - this completes quickly
2. **But save doesn't complete** until `refresh_entry().await?` finishes
3. **refresh_entry** sends a scan request to a background thread and waits for it
4. **The background scanner** does filesystem operations to update Zed's internal state
5. **Only then** does the save operation return success

### Impact on Hot Reload

- **Other editors** (VS Code, Sublime) return from save immediately after writing to disk
- **Zed blocks** the save completion until its internal filesystem state is refreshed
- **Hot reload tools** typically watch for:
  - File modification events from the OS
  - Editor save completion signals
- **The delay** means hot reload sees the file change but Zed appears "still saving"

## Reproduction Steps

1. Set up a project with hot reload (e.g., Vite, Next.js dev server)
2. Open the same file in both Zed and VS Code/Sublime
3. Make changes and save in both editors
4. Observe that hot reload triggers faster with VS Code/Sublime than with Zed

## Proposed Solution

Make the `refresh_entry` call non-blocking so the save operation completes immediately after the file write, with refresh happening asynchronously in the background.

**Current flow:**
```
Write File → Wait for Refresh → Return Success
```

**Should be:**
```
Write File → Return Success
           ↓
    Background Refresh
```

## Implementation

The fix involves modifying the `write_file` function to:

1. Start the refresh task but don't await it
2. Spawn refresh in background with `.detach()`
3. Return success immediately after file write
4. Try to use existing entry state when possible

```rust
cx.spawn(async move |this, cx| {
    write.await?;
    
    // Start refresh asynchronously but don't wait for it to complete
    let refresh_task = this.update(cx, |this, cx| {
        this.as_local_mut()
            .unwrap()
            .refresh_entry(path.clone(), None, cx)
    })?;
    
    // Spawn refresh in background - don't block save completion
    cx.background_spawn(async move {
        refresh_task.await.log_err();
    })
    .detach();
    
    let worktree = this.upgrade().context("worktree dropped")?;
    
    // Try to get existing entry first
    let entry = this.read_with(cx, |this, _| {
        this.entry_for_path(&path).cloned()
    })?;
    
    if let Some(entry) = entry {
        Ok(File::for_entry(entry, worktree))
    } else {
        // If no entry exists, create a minimal File representation
        let metadata = fs.metadata(&abs_path).await?;
        Ok(Arc::new(File {
            worktree,
            path,
            disk_state: DiskState::Present { mtime: metadata.mtime },
            entry_id: None,
            is_local: true,
            is_private,
        }))
    }
})
```

## Benefits

1. **Faster hot reload**: Save completion happens immediately after file write
2. **Better dev experience**: Consistent with other popular editors
3. **No functionality loss**: Filesystem state still gets updated asynchronously
4. **Maintains correctness**: File metadata is still available for immediate use

## Testing

To verify the fix:

1. Set up a simple hot reload project
2. Time the delay between save and hot reload trigger
3. Compare before/after the fix
4. Test with various project types (Vite, Next.js, etc.)

## Files Changed

- `crates/worktree/src/worktree.rs` - Modified `LocalWorktree::write_file` implementation

## Environment

- **OS**: macOS (likely affects other platforms too)
- **Zed Version**: Current main branch
- **Hot Reload Tools**: Vite, Next.js dev server, and others

**Comment by sausaw** [comment]
Pull Request Created: https://github.com/zed-industries/zed/pull/34529

The fix modifies the write_file function to make refresh_entry non-blocking, allowing save operations to complete immediately after file write while filesystem state refresh happens asynchronously in the background.

**Comment by MrSubidubi** [comment]
You completely skipped/removed the issue template whilst filing this issue and it's very hard for us to both debug and reproduce. Please open another issue where you are using the template or edit this issue to be more concise so it is easier for us to take a look. Thanks
